src/render/graph.rs

```diff
@@ -180,13 +183,32 @@ impl Graph {
         marked_temp: &mut Vec<NodeIndex>,
         ordered: &mut Vec<NodeIndex>,
         in_cycle: &mut Vec<NodeIndex>,
+        cycle_breakers: &mut Vec<NodeIndex>,
     ) {
         // If this node is in the cycle detection list, it is part of a cycle!
         if let Some(pos) = marked_temp.iter().position(|&m| m == node_id) {
-            // Mark all nodes in the cycle
-            in_cycle.extend_from_slice(&marked_temp[pos..]);
-            // Do not continue, as we already have visited all these nodes
-            return;
+            // check if we can find some node that can break the cycle
+            let cycle_breaker_node = marked_temp.iter().skip(pos).find(|node_id| {
+                let node = self.nodes.get(node_id).unwrap();
+                node.processor.can_break_cycle()
+            });
+
+            match cycle_breaker_node {
+                Some(&node_id) => {
+                    // store node id to clear the node outgoing edges
+                    cycle_breakers.push(node_id);
+                    // remove nodes from mark temp after pos
+                    marked_temp.truncate(pos);
```

Comment:
I'm not completely sure of this

Reply:
Yeah, it's quite hard to reason about it. My gut feeling is that this implementation is to optimistic. It would be safer to clear out `marked_temp` and also remove these nodes from `marked`. Then start visiting again from the first item that was in `marked_temp`

Reply:
Maybe a nice thing to have would be a unit test with really weird edge case: some cycle inside a cycle I guess, to check this behaves as expected (whatever the solution)?